AsyncDisplayKit/ASCollectionNode.h

```diff
@@ -131,20 +241,346 @@ NS_ASSUME_NONNULL_BEGIN
+ *
+ * @return a block that creates the node for display at this indexpath.
+ *   Must be thread-safe (can be called on the main thread or a background
+ *   queue) and should not implement reuse (it will be called once per row).
+ */
+- (ASCellNodeBlock)collectionView:(ASCollectionView *)collectionView nodeBlockForItemAtIndexPath:(NSIndexPath *)indexPath ASDISPLAYNODE_DEPRECATED;
+
+/**
+ * Asks the collection view to provide a supplementary node to display in the collection view.
+ *
+ * @param collectionView An object representing the collection view requesting this information.
+ * @param kind           The kind of supplementary node to provide.
+ * @param indexPath      The index path that specifies the location of the new supplementary node.
+ */
+- (ASCellNode *)collectionView:(ASCollectionView *)collectionView nodeForSupplementaryElementOfKind:(NSString *)kind atIndexPath:(NSIndexPath *)indexPath ASDISPLAYNODE_DEPRECATED;
+
+/**
+ * Indicator to lock the data source for data fetching in async mode.
+ * We should not update the data source until the data source has been unlocked. Otherwise, it will incur data inconsistency or exception
+ * due to the data access in async mode.
+ *
+ * @param collectionView The sender.
+ * @deprecated The data source is always accessed on the main thread, and this method will not be called.
+ */
+- (void)collectionViewLockDataSource:(ASCollectionView *)collectionView ASDISPLAYNODE_DEPRECATED;
+
+/**
+ * Indicator to unlock the data source for data fetching in async mode.
+ * We should not update the data source until the data source has been unlocked. Otherwise, it will incur data inconsistency or exception
+ * due to the data access in async mode.
+ *
+ * @param collectionView The sender.
+ * @deprecated The data source is always accessed on the main thread, and this method will not be called.
+ */
+- (void)collectionViewUnlockDataSource:(ASCollectionView *)collectionView ASDISPLAYNODE_DEPRECATED;
+
+@end
+
+/**
+ * This is a node-based UICollectionViewDelegate.
+ */
+@protocol ASCollectionDelegate <ASCommonCollectionDelegate, NSObject>
+
+@optional
+
+/**
+ * Provides the constrained size range for measuring the given item.
+ *
+ * @param collectionNode The sender.
+ *
+ * @param indexPath The index path of the item.
+ *
+ * @return A constrained size range for layout for the item at this index path.
+ */
+- (ASSizeRange)collectionNode:(ASCollectionNode *)collectionNode constrainedSizeForItemAtIndexPath:(NSIndexPath *)indexPath;
+
+- (void)collectionNode:(ASCollectionNode *)collectionNode willDisplayItemWithNode:(ASCellNode *)node;
+
+- (void)collectionNode:(ASCollectionNode *)collectionNode didEndDisplayingItemWithNode:(ASCellNode *)node;
+
+- (BOOL)collectionNode:(ASCollectionNode *)collectionNode shouldHighlightItemWithNode:(ASCellNode *)node;
+- (void)collectionNode:(ASCollectionNode *)collectionNode didHighlightItemWithNode:(ASCellNode *)node;
```

Comment:
Well, I've reversed stance on this. In order to keep things simple, all delegate callbacks except `willDisplayRow` and `didEndDisplayingRow` will pass a `nonnull NSIndexPath *`. In the extremely unlikely case that the user selected an item that is pending deletion/reload, the event won't get sent.